[Linux] Web Inspector: add memory usage of images in memory timeline

pgorszkowski-igalia opened the pull request:
#### f3204eb2dfb23d159bedc68378578043b1c21d16
<pre>
[Linux] Web Inspector: add memory usage of images in memory timeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=253873">https://bugs.webkit.org/show_bug.cgi?id=253873</a>

Reviewed by Carlos Garcia Campos.

This fix extends what is reported in memory timeline(web inspector) of
memory usage with data for images(only for linux).

* Source/WebCore/page/ResourceUsageThread.cpp:
(WebCore::ResourceUsageThread::setTotalLayerInfo):
* Source/WebCore/page/ResourceUsageThread.h:
* Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
(WebCore::ResourceUsageThread::platformCollectMemoryData):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingLayers):

Canonical link: <a href="https://commits.webkit.org/263230@main">https://commits.webkit.org/263230@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee3c0824c20620823a2a4df30a7a83b7097bf602

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3544 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3591 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3729 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4967 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3818 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3522 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3696 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3632 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3077 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3586 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3851 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3191 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4789 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1358 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3167 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3067 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3158 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3226 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4555 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3602 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2908 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3144 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3165 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/975 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3170 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3422 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shivamidow commented on Source/WebCore/page/ResourceUsageThread.cpp:
```diff
@@ -77,6 +77,13 @@ void ResourceUsageThread::removeObserver(void* key)
+{
+    auto& resourceUsageThread = ResourceUsageThread::singleton();
+
+    resourceUsageThread.totalLayerBackingStoreBytes = layerBackingStoreBytes;
```
/* Unofficial Review */
We can simplify this method.
`ResourceUsageThread::singleton().totalLayerBackingStoreBytes = layerBackingStoreBytes;`

By the way, an explicit method name is preferred. The only 'info' here is the total backing store size.
Besides, this method seems not thread-safe. You may need a lock.

pgorszkowski-igalia replied:
Thanks for simplified version.
This not thread-safe issue also was in my mind. I will add a lock for that.

shivamidow commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -312,10 +313,15 @@ void ResourceUsageThread::platformCollectMemoryData(JSC::VM* vm, ResourceUsageDa
-    size_t categoriesTotalSize = 0;
-    for (auto& category : data.categories)
-        categoriesTotalSize += category.totalSize();
-    data.categories[MemoryCategory::Other].dirtySize = data.totalDirtySize - categoriesTotalSize;
```
/* Unofficial Review */
Does the MemoryCategory::Other consist of MemoryCategory::Images, MemoryCategory::LibcMalloc, and MemoryCategory::Layers? Isn't there any other resource in MemoryCategory::Other, except for them?

pgorszkowski-igalia replied:
I was thinking about that and I was considering to leave it. MemoryCategory::Other will contain what is not in other categories(JS, images, page, layers).

pgorszkowski-igalia replied:
data.totalDirtySize is calculated base on data from /proc/self/statm (resident-share):
resident   (2) resident set size
                             (inaccurate; same as VmRSS in /proc/[pid]/status)
shared     (3) number of resident shared pages
                             (i.e., backed by a file)
                             (inaccurate; same as RssFile+RssShmem in
                             /proc/[pid]/status)

while in my change LibcMalloc category is:
data.categories[MemoryCategory::LibcMalloc].dirtySize = WTF::fastMallocStatistics().committedVMBytes;
It is taken from "getrusage" (ru_maxrss) - https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/FastMalloc.cpp#L677
ru_maxrss;        /* maximum resident set size */ 

I have checked that in case of Win and Mac, the LibcMalloc category is calculated in different way, but IMO it could be reused also in case of Linux:
(Mac): https://github.com/WebKit/WebKit/blob/main/Source/WebCore/page/cocoa/ResourceUsageThreadCocoa.mm#L234
(Win): https://github.com/WebKit/WebKit/blob/main/Source/WebCore/page/win/ResourceUsageThreadWin.cpp#L133

@shivamidow : what do you think?

pgorszkowski-igalia replied:
At the end I decided to remove what was collected in LibcMalloc category. In inspector "page" and "other" are displayed as one category(https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Models/MemoryTimelineRecord.js#L72). So I leave how "other" category is calculated.

carlosgcampos replied:
I'm not sure I understand this discussion, but I do't understand why the other category is removed, it's supposed to be the rest of the memory not included in categories.

shivamidow commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -950,6 +951,11 @@ bool RenderLayerCompositor::updateCompositingLayers(CompositingUpdateType update
             m_secondaryCompositedLayerCount, m_obligatoryBackingStoreBytes / 1024, m_secondaryBackingStoreBytes / 1024, (m_obligatoryBackingStoreBytes + m_secondaryBackingStoreBytes) / 1024, (endTime - startTime).milliseconds());
+
+#if ENABLE(RESOURCE_USAGE)
+        // set the total layer memory into ResourceUsageThread for Inspector layers memorytimeline.
```
/* Unofficial Review */
// We save the total memory size of layers to show in the memory timeline of Inspector.
?

carlosgcampos reviewed (commented):
Images size looks good, but before adding support for layers sizes we should probably implement GraphicsLayer::backingStoreMemoryEstimate() for texture mapper.

carlosgcampos commented on Source/WebCore/page/ResourceUsageThread.h:
```diff
@@ -80,6 +80,7 @@ class ResourceUsageThread {
 
     RefPtr<Thread> m_thread;
     Lock m_observersLock;
+    Lock m_totalLayerBackingStoreLock;
```
This is not need for images only.

carlosgcampos commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -312,11 +313,18 @@ void ResourceUsageThread::platformCollectMemoryData(JSC::VM* vm, ResourceUsageDa
         categoriesTotalSize += category.totalSize();
-    data.categories[MemoryCategory::Other].dirtySize = data.totalDirtySize - categoriesTotalSize;
 
+    data.categories[MemoryCategory::Other].dirtySize = data.totalDirtySize - categoriesTotalSize;
```
Ah, I see it's here again. What changed then?

pgorszkowski-igalia replied:
It stays as it was: "other" category is calculated base on currentProcessMemoryStatus(memoryStatus.resident - memoryStatus.shared) minus what is in other categories(JS, Images currently)

carlosgcampos commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -57,6 +57,7 @@
 #include "RenderLayerScrollableArea.h"
 #include "RenderVideo.h"
 #include "RenderView.h"
+#include "ResourceUsageThread.h"
```
This shouldn't be needed in this patch either.

pgorszkowski-igalia committed:
[Linux] Web Inspector: add memory usage of images in memory timeline
https://bugs.webkit.org/show_bug.cgi?id=253873

Reviewed by Carlos Garcia Campos.

This fix extends what is reported in memory timeline(web inspector) of
memory usage with data for images(only for linux).

* Source/WebCore/page/ResourceUsageThread.cpp:
(WebCore::ResourceUsageThread::setTotalLayerInfo):
* Source/WebCore/page/ResourceUsageThread.h:
* Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
(WebCore::ResourceUsageThread::platformCollectMemoryData):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingLayers):

Canonical link: https://commits.webkit.org/263230@main
